### PR TITLE
[docs] add pages about examples on training language models with fairseq

### PR DESCRIPTION
**Pull request by zhuohan123**
## Why are these changes needed?

Provide a new example to use Ray for fault-tolerant training of language models on the cloud.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.

### CODE REVIEW

File: doc/examples/lm/ray_train.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+TOTAL_UPDATES=125000       # Total number of training steps
```
Comment (richardliaw) [inline doc/examples/lm/ray_train.sh]:
```suggestion
#!/bin/bash

TOTAL_UPDATES=125000       # Total number of training steps
```

File: doc/source/example-lm.rst [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+                  finished = ray.get(finished)
+              retry = False
+          except Exception as inst:
+              print("Ray restart because following error occurs:")
+              print(inst)
+              retry = True
+          ray.shutdown()
+
+
+  if __name__ == '__main__':
```
Comment (richardliaw) [inline doc/source/example-lm.rst]: you can remove this

File: doc/source/example-lm.rst [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+          try:
+              while len(unfinished) > 0:
+                  finished, unfinished = ray.wait(unfinished)
+                  finished = ray.get(finished)
+              retry = False
+          except Exception as inst:
+              print("Ray restart because following error occurs:")
+              print(inst)
+              retry = True
+          ray.shutdown()
```
Comment (richardliaw) [inline doc/source/example-lm.rst]: Can you make this into a separate code block with a caption on top?

File: doc/source/example-lm.rst [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                        help='address for ray initialization')
+      group.add_argument('--fix-batch-size', default=None, type=int,
+                        help='fix batch size (max_sentences * update_freq '
+                              '* n_GPUs) to be a fixed input value for different '
+                              'number of GPUs or CPUs')
+      # fmt: on
+      return group
+
+
+  def ray_main():
```
Comment (richardliaw) [inline doc/source/example-lm.rst]: can you change this to "run_fault_tolerant_loop"?

File: doc/source/example-lm.rst [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+          else:
+              original_n_gpus = args.distributed_world_size
+
+              def _new_save_checkpoint(*args, **kwargs):
+                  _original_save_checkpoint(*args, **kwargs)
+                  n_gpus = int(ray.cluster_resources().get("GPU", 0))
+                  if n_gpus > original_n_gpus:
+                      raise Exception("New GPUs find (original %d GPUs, now %d GPUs)"
+                                      % (original_n_gpus, n_gpus))
+          fairseq.checkpoint_utils.save_checkpoint = _new_save_checkpoint
```
Comment (richardliaw) [inline doc/source/example-lm.rst]: Can you put this into a helper function that passes in args?

Then, we can add the comment as a separate code block.

File: doc/source/example-lm.rst (hunk 3 above)
Comment (zhuohan123) [reply doc/source/example-lm.rst]: resolved in the latest comment

File: doc/examples/lm/ray_train.sh (hunk 1 above)
Comment (zhuohan123) [reply doc/examples/lm/ray_train.sh]: resolved in the latest comment

File: doc/source/example-lm.rst (hunk 4 above)
Comment (zhuohan123) [reply doc/source/example-lm.rst]: resolved in the latest comment

File: doc/source/example-lm.rst (hunk 5 above)
Comment (zhuohan123) [reply doc/source/example-lm.rst]: resolved in the latest comment

File: doc/examples/lm/lm-cluster.yaml [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+# Provider-specific config for the head node, e.g. instance type. By default
+# Ray will auto-configure unspecified fields such as SubnetId and KeyName.
+# For more documentation on available fields, see:
+# http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.ServiceResource.create_instances
+head_node:
+    InstanceType: m5.xlarge
+    ImageId: ami-0b294f219d14e6a82 # Deep Learning AMI (Ubuntu) Version 21.0
+    SecurityGroupIds:
+        - "{{SecurityGroupId}}"
```
Comment (richardliaw) [inline doc/examples/lm/lm-cluster.yaml]: we don't need this right?

Comment (richardliaw) [reply doc/examples/lm/lm-cluster.yaml]: jk, it's for efs

**Comment by richardliaw** [comment]
@zhuohan123 this is very close!

I ran this and ran into issues - 

1. Error
```
2019-10-20 04:23:43,290 ERROR worker.py:1719 -- Possible unhandled error from worker: ray_worker (pid=3966, host=ip-172-31-32-139)
  File "/home/ubuntu/efs/ray_train.py", line 40, in run
    main(args, init_distributed=(args.distributed_world_size > 1))
  File "/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/fairseq_cli/train.py", line 68, in main
    extra_state, epoch_itr = checkpoint_utils.load_checkpoint(args, trainer)
  File "/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/fairseq/checkpoint_utils.py", line 128, in load_checkpoint
    trainer.lr_step(epoch_itr.epoch)
  File "/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/fairseq/trainer.py", line 464, in lr_step
    self.lr_scheduler.step(epoch, val_loss)
  File "/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/fairseq/trainer.py", line 98, in lr_scheduler
    self._build_optimizer()  # this will initialize self._lr_scheduler
  File "/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/fairseq/trainer.py", line 110, in _build_optimizer
    self._optimizer = optim.FP16Optimizer.build_optimizer(self.args, params)
  File "/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/fairseq/optim/fp16_optimizer.py", line 103, in build_optimizer
    return cls(args, params, fp32_optimizer, fp32_params)
  File "/home/ubuntu/anaconda3/envs/pytorch_p36/lib/python3.6/site-packages/fairseq/optim/fp16_optimizer.py", line 68, in __init__
    if len(args.update_freq) > 1:
TypeError: object of type 'int' has no len()
```

2. It wasn't clear to me that the files in `lm/` had to go to `$HOME/efs`. I've added a line in the docs for this.

**Comment by zhuohan123** [comment]
@richardliaw Thanks for the detailed reviews! I have fixed the bug and unified all the paths involved to `$HOME/efs/lm`

**Comment by richardliaw** [comment]
```
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.50 GiB (GPU 0; 15.75 GiB total capacity; 11.71 GiB already allocated; 584.94 MiB free; 2.38 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.36 GiB (GPU 0; 15.75 GiB total capacity; 11.64 GiB already allocated; 584.94 MiB free; 2.45 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.52 GiB (GPU 0; 15.75 GiB total capacity; 12.18 GiB already allocated; 584.94 MiB free; 1.91 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.47 GiB (GPU 0; 15.75 GiB total capacity; 11.62 GiB already allocated; 584.94 MiB free; 2.47 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.50 GiB (GPU 0; 15.75 GiB total capacity; 11.92 GiB already allocated; 584.94 MiB free; 2.17 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.52 GiB (GPU 0; 15.75 GiB total capacity; 12.11 GiB already allocated; 584.94 MiB free; 1.98 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.53 GiB (GPU 0; 15.75 GiB total capacity; 12.05 GiB already allocated; 584.94 MiB free; 2.04 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.39 GiB (GPU 0; 15.75 GiB total capacity; 12.04 GiB already allocated; 584.94 MiB free; 2.05 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.41 GiB (GPU 0; 15.75 GiB total capacity; 12.50 GiB already allocated; 584.94 MiB free; 1.59 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.36 GiB (GPU 0; 15.75 GiB total capacity; 11.93 GiB already allocated; 584.94 MiB free; 2.16 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.40 GiB (GPU 0; 15.75 GiB total capacity; 12.31 GiB already allocated; 584.94 MiB free; 1.78 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.42 GiB (GPU 0; 15.75 GiB total capacity; 12.33 GiB already allocated; 584.94 MiB free; 1.76 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.38 GiB (GPU 0; 15.75 GiB total capacity; 12.14 GiB already allocated; 584.94 MiB free; 1.95 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.53 GiB (GPU 0; 15.75 GiB total capacity; 12.22 GiB already allocated; 584.94 MiB free; 1.87 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.39 GiB (GPU 0; 15.75 GiB total capacity; 11.99 GiB already allocated; 584.94 MiB free; 2.10 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.42 GiB (GPU 0; 15.75 GiB total capacity; 12.33 GiB already allocated; 584.94 MiB free; 1.76 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.39 GiB (GPU 0; 15.75 GiB total capacity; 12.27 GiB already allocated; 584.94 MiB free; 1.82 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.50 GiB (GPU 0; 15.75 GiB total capacity; 11.65 GiB already allocated; 584.94 MiB free; 2.44 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
(pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.53 GiB (GPU 0; 15.75 GiB total capacity; 12.05 GiB already allocated; 584.94 MiB free; 2.04 GiB cached);
(pid=3959, ip=172.31.41.219)  Skipping batch
```

**Comment by zhuohan123** [comment]
> ```
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.50 GiB (GPU 0; 15.75 GiB total capacity; 11.71 GiB already allocated; 584.94 MiB free; 2.38 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.36 GiB (GPU 0; 15.75 GiB total capacity; 11.64 GiB already allocated; 584.94 MiB free; 2.45 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.52 GiB (GPU 0; 15.75 GiB total capacity; 12.18 GiB already allocated; 584.94 MiB free; 1.91 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.47 GiB (GPU 0; 15.75 GiB total capacity; 11.62 GiB already allocated; 584.94 MiB free; 2.47 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.50 GiB (GPU 0; 15.75 GiB total capacity; 11.92 GiB already allocated; 584.94 MiB free; 2.17 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.52 GiB (GPU 0; 15.75 GiB total capacity; 12.11 GiB already allocated; 584.94 MiB free; 1.98 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.53 GiB (GPU 0; 15.75 GiB total capacity; 12.05 GiB already allocated; 584.94 MiB free; 2.04 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.39 GiB (GPU 0; 15.75 GiB total capacity; 12.04 GiB already allocated; 584.94 MiB free; 2.05 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.41 GiB (GPU 0; 15.75 GiB total capacity; 12.50 GiB already allocated; 584.94 MiB free; 1.59 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.36 GiB (GPU 0; 15.75 GiB total capacity; 11.93 GiB already allocated; 584.94 MiB free; 2.16 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.40 GiB (GPU 0; 15.75 GiB total capacity; 12.31 GiB already allocated; 584.94 MiB free; 1.78 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.42 GiB (GPU 0; 15.75 GiB total capacity; 12.33 GiB already allocated; 584.94 MiB free; 1.76 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.38 GiB (GPU 0; 15.75 GiB total capacity; 12.14 GiB already allocated; 584.94 MiB free; 1.95 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.53 GiB (GPU 0; 15.75 GiB total capacity; 12.22 GiB already allocated; 584.94 MiB free; 1.87 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.39 GiB (GPU 0; 15.75 GiB total capacity; 11.99 GiB already allocated; 584.94 MiB free; 2.10 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.42 GiB (GPU 0; 15.75 GiB total capacity; 12.33 GiB already allocated; 584.94 MiB free; 1.76 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.39 GiB (GPU 0; 15.75 GiB total capacity; 12.27 GiB already allocated; 584.94 MiB free; 1.82 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.50 GiB (GPU 0; 15.75 GiB total capacity; 11.65 GiB already allocated; 584.94 MiB free; 2.44 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> (pid=3959, ip=172.31.41.219) | WARNING: ran out of memory with exception: CUDA out of memory. Tried to allocate 1.53 GiB (GPU 0; 15.75 GiB total capacity; 12.05 GiB already allocated; 584.94 MiB free; 2.04 GiB cached);
> (pid=3959, ip=172.31.41.219)  Skipping batch
> ```

Sorry I forgot to push the latest commit. I did a test run just now and it trains smoothly.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**